File: exercise/scala/day18/src/test/scala/RockPaperScissorsSteps.scala
import game.{Choice, Result, RockPaperScissors, Winner}
import io.cucumber.scala.{EN, ScalaDsl}
import org.scalatest.matchers.should.Matchers

class RockPaperScissorsSteps extends ScalaDsl with EN with Matchers {
  private var result: Result = _
  private var player1: Choice = _
  private var player2: Choice = _

  private def parseChoice(choice: String): Choice = choice match {
    case "🪨" => Choice.ROCK
    case "📄" => Choice.PAPER
    case "✂️" => Choice.SCISSORS
    case _ => throw new IllegalArgumentException("Invalid choice")
  }

  private def parseWinner(winner: String): Winner = winner match {
    case "Player 1" => Winner.PLAYER_1
    case "Player 2" => Winner.PLAYER_2
    case "Draw" => Winner.DRAW
    case _ => throw new IllegalArgumentException("Invalid winner")
  }

  Given("""^Player (\d+) chooses (.*)$""") { (player: Int, choice: String) =>
    val selectedChoice = parseChoice(choice)
    if (player == 1) player1 = selectedChoice else player2 = selectedChoice
  }

  When("""^they play$""") { () =>
    result = RockPaperScissors.play(player1, player2)
  }

  Then("""^the result should be (.*) because (.*)$""") { (expectedWinner: String, expectedReason: String) =>
    result.winner shouldEqual parseWinner(expectedWinner)
    result.reason shouldEqual expectedReason
  }
}

File: exercise/kotlin/day18/src/test/java/RockPaperScissorsSteps.java
import game.Choice;
import game.Result;
import game.RockPaperScissors;
import game.Winner;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import static org.assertj.core.api.Assertions.assertThat;

public class RockPaperScissorsSteps {
    private Result result;
    private Choice player1;
    private Choice player2;

    private static Choice parseChoice(String choice) {
        return switch (choice) {
            case "🪨" -> Choice.ROCK;
            case "📄" -> Choice.PAPER;
            case "✂️" -> Choice.SCISSORS;
            default -> throw new IllegalArgumentException("Invalid choice");
        };
    }

    private static Winner parseWinner(String winner) {
        return switch (winner) {
            case "Player 1" -> Winner.PLAYER_1;
            case "Player 2" -> Winner.PLAYER_2;
            case "Draw" -> Winner.DRAW;
            default -> throw new IllegalArgumentException("Invalid winner");
        };
    }

    @Given("^Player (\\d+) chooses (.*)$")
    public void playerChooses(int player, String choice) {
        var selectedChoice = parseChoice(choice);

        if (player == 1) player1 = selectedChoice;
        else player2 = selectedChoice;
    }

    @When("^they play$")
    public void theyPlay() {
        result = RockPaperScissors.Companion.play(player1, player2);
    }

    @Then("^the result should be (.*) because (.*)$")
    public void theResultShouldBe(String expectedWinner, String expectedReason) {
        assertThat(result.getWinner()).isEqualTo(parseWinner(expectedWinner));
        assertThat(result.getReason()).isEqualTo(expectedReason);
    }
}